Construct NJT without graph breaks

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132414
* __->__ #130292
* #131803
* #131916


Combines contributions from https://github.com/pytorch/pytorch/pull/130505

Some context can be found in this large comment block:

https://github.com/pytorch/pytorch/blob/a5b64d39fd50d47dfe6f821941ce4c803c7c4d63/test/dynamo/test_subclasses.py#L1667-L1681

Changes in this PR 
- For each tensor fakified, check the nested int registry in eager, and eagerly symbolicize if that tensor has already been associated with nested int in eager.
- Adds a separate counter stored on FakeTensorMode as a fake analog to _tensor_id_counter (which keeps track of unique tensors). This counter is initialized to the global eager tensor id counter upon creation of the FakeTensorMode, and needs to be reset when the same FakeTensorMode is reused to trace again (in this PR, we piggyback on the epoch incrementing logic).
- (refactor) Today, we store FakeTensor -> symbolic nested int in the global registry. With this PR, symbolic nested int is stored directly on the FakeTensor. (Eager still caches nested int in the registry, though we should avoid this at some point.)

Basically unchanged, but worth noting:
- `__tensor_unflatten__` is still responsible for determining whether we should cache for now. The logic is somewhat simplified.
- to_copy is still using the trick of updating two different tensors in the registry to point to the same nested int. This is kind of broken, but we try to leave it as is, and plan a better fix with the UnionFind stack.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

Differential Revision: [D60406772](https://our.internmc.facebook.com/intern/diff/D60406772)

soulitzer committed:
Construct NJT without graph breaks

[ghstack-poisoned]

soulitzer commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -160,7 +160,10 @@ def inner(*flat_args):
+
+            tmp_counter = torch.nested._internal.nested_tensor._tensor_id_counter
             flat_f_outs = f(*flat_f_args)
+            torch.nested._internal.nested_tensor._tensor_id_counter = tmp_counter
```
Maybe there's a more principled way to do this

soulitzer replied:
Without this change, the case where NJT is created from offsets which itself is an intermediate tensor would break.

It should be breaking because (1) aot autograd traces forward twice, one to record some aliasing metadata, and a second time to trace out the joint graph, (2) aot autograd creates tangents from the output of the first forward to use as inputs to the joint graph.

Today when we trace, we mutate the global state, meaning that if we try to trace multiple times we produces different results, leading to a mismatch in shapes.

(Also what we may want to do in addition to resetting is to having a separate counter during tracing, so that the eager state is not affected.)

> maybe a context manager?

Yeah, this may be better if we want to stick with this solution, but I was lazy here lol

soulitzer committed:
Update on "Construct NJT without graph breaks"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

jbschlosser commented on test/dynamo/test_subclasses.py:
```diff
@@ -1507,6 +1507,95 @@ def binary(nt1, nt2):
+        self.assertEqual(out, ref_out)
+        self.assertEqual(grad, ref_grad)
+
+        #
+        # Mixed case (This case still fails!)
+        #
+        # values = nt.values().detach() # requires_grad_(False)
+        # torch._dynamo.mark_dynamic(values, 0)
+
+        # def fn(nt, values, offsets):
+        #     nt2 = torch.nested.nested_tensor_from_jagged(values, offsets)
+        #     return nt * nt2
+
+        # out = torch.compile(fn, fullgraph=True, backend="aot_eager")(nt, values, nt.offsets())
```
curious what gaps are left that make this case still fail? I'm playing around with it locally and it's tricky to nail down

jbschlosser commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -622,13 +627,17 @@ def inner(*flat_args):
             if isinstance(t, Tensor):
-                return t.view(t.shape)
+                out = t.view(t.shape)
+                if parent_cls is  torch.nested._internal.nested_tensor.NestedTensor and (attr == "_offsets" or attr == "_lengths"):
```
I'm not crazy about this NJT-specific logic here but I don't have an alternative suggestion yet - working on this though

jbschlosser replied:
slightly better approach in #130505 maybe?

jbschlosser reviewed (commented):
Nice work! Looking forward to putting this graph break during construction stuff behind us forever :)

I've got some questions throughout but my understanding is that the key change here is to expand usage of the offsets -> nested int registry to better support functional / fake tensors.
* Functional tensors: simply unwrap the underlying non-functional tensor and use that to query the registry.
* Fake tensors: operate on symbolic nested ints with valid sources. @davidberard98 this avoids the need for an `is_concrete_int()` change like #124803, since, after this PR, we'll hit that logic with a symbolic nested int wrapping a normal nested int, which does have a `.node.expr`.

I believe there's also some logic to work around an AOTAutograd detail (i.e. `Views to avoid tangents aliasing inputs`) by manually populating the offsets -> nested int registry for the view tensor.

@soulitzer any high-level context I'm missing or have wrong?

soulitzer commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -557,6 +557,10 @@ class FakeTensor(torch.Tensor):
 
+    # This is set when an tensor input is fakified. It is used when constructing
+    # a NestedIntSource from an offsets Tensor.
+    source = None
```
It is set for every tensor actually! Which is not great, but hard to avoid because during fakification, we don't know yet whether this tensor will be used in NT construction yet.

soulitzer commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -16,8 +16,37 @@ def get_tensor_symint(tensor, *, coeff=1):
     if tensor_symint is None:
-        tensor_symint = torch._C._get_nested_int(_tensor_id_counter, coeff)
-        _tensor_id_counter += 1
+        from torch._subclasses.fake_tensor import FakeTensor
```
Sg!

soulitzer commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -16,8 +16,37 @@ def get_tensor_symint(tensor, *, coeff=1):
+                src = torch._dynamo.source.EphemeralSource("intermediate_offsets")
+            else:
+                src = torch._dynamo.source.NestedIntSource(tensor.source)
+            ret = shape_env.create_symintnode(
+                sym=shape_env.create_symbol(
+                    val=hint,
+                    source=src,
+                ),
+                hint=hint,
+                source=src,
+            )
```
Good find. Yep, this should be corrected.

soulitzer commented on torch/_subclasses/meta_utils.py:
```diff
@@ -1545,6 +1545,7 @@ def is_c_of_r(complex_dtype, real_dtype):
                 r._is_param = True
 
             self.set_tensor_memo(t, r)
+            r.source = source
```
Yeah

soulitzer commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3960,6 +3960,10 @@ def track_symfloat(source, val):
             else:
                 sources_tensors_constraints = [(source, t, context.constraint_sizes)]
 
+                from torch.nested._internal.nested_tensor import _tensor_symint_registry
+                if t in _tensor_symint_registry:
+                    track_symint(torch._dynamo.source.NestedIntSource(source), _tensor_symint_registry[t])
```
Part of the reason I'm doing it this way vs. unbacked symints is that unbacked symint approach does not work (it worked at one point). I'm still working through those issues.

It may not necessary for this PR actually, but two cases where we eventually need to have a non-unbacked symint is:

(1) In this case, offset1 and offset2 are not the same tensor object, so we cannot rely on dynamo graph arg deduping guards to trigger. 
```
trust_me_assert_equal(offset1, offset2)

def fn(values1, offset1, values2, offset2):
    nt1 = make_nt(values1, offset1)
    nt2 = make_nt(values2, offset2)
    ...
```

(2) Factory functions

soulitzer commented on torch/_dynamo/source.py:
```diff
@@ -285,6 +285,25 @@ def name(self):
+    def __post_init__(self):
+        assert self.base is not None
+
+    def reconstruct(self, codegen):
+        # This is wrong, when is this used?
+        return [
+            *self.base.reconstruct(codegen),
+            codegen.create_load_attr(self.prop),
+        ]
```
Just going to follow the other ChainedSource examples in this file for now.

soulitzer committed:
Update on "Construct NJT without graph breaks"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "Construct NJT without graph breaks"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1105,6 +1165,7 @@ def __exit__(self, a, b, c):
                 torch._C._set_dispatch_mode(maybe_prev_fake_mode)
             if maybe_prev_only_lift_cpu_tensors is not None:
                 torch._C._set_only_lift_cpu_tensors(maybe_prev_only_lift_cpu_tensors)
+            self.nt_tensor_id_counter = self.nt_tensor_id_initial_count
```
This assumes that we don't enter/exit multiple times during a tracing mode. Need to actually check this.

jbschlosser reviewed (commented):
btw you may have seen this interesting test case involving multiple ShapeEnvs:
```
PYTORCH_TEST_WITH_DYNAMO=1 python test/test_fake_tensor.py -k FakeTensorTest.test_jagged_fake_to_fake_preserved
```

I had to add some logic to enforce the invariant that any `fake offsets -> symbolic nested int` entry in the registry has consistent ShapeEnv between the fake offsets and symbolic nested int. But maybe you've sidestepped this via your `FakeTensor` / `FakeTensorMode` changes?

Some questions / comments:
> (1) Symbolic nested int are now cached on FakeTensor

Does this mean that the nested int registry only ever holds `real tensor -> non-symbolic nested int` entries now?

> (2) It is now the responsibility of the caller of __tensor_unflatten__ to properly update the symbolic nested int cache.

I'm a little worried about the completeness of this / how easy it is to ensure this is being done correctly. I think I'd argue that `__tensor_unflatten__()` is a little late and it's nicer to ensure the symbolic nested int cache is updated immediately upon creation of a new (fake, functional, view) tensor that conceptually has the same ragged structure as a given `offsets` tensor. This is a bit stronger invariant and we know downstream when `__tensor_unflatten__()` is eventually called everything will be okay.

> (3) New trace-time specific tensor id counter. (This needs more testing)

This I do like :) much nicer than hackery of restoring the counter during fw graph invocation for metadata collection.

> (4) Source is stored on FakeTensor for ALL tensors upon fakification

not crazy about this but I don't see a way around it IF we ever need to get a real source / guard on a nested int. I think we should either boot this or actually make sure the tests in this PR test cases like this.

soulitzer commented:
> Does this mean that the nested int registry only ever holds  real tensor -> non-symbolic nested int entries now?

Yeah

>  think I'd argue that __tensor_unflatten__() is a little late and it's nicer to ensure the symbolic nested int cache is updated immediately upon creation of a new (fake, functional, view) tensor that conceptually has the same ragged structure as a given offsets tensor.

Agreed! I think I mainly meant that `__tensor_unflatten__` would no longer be responsible, and yes upon creation sounds like the preferred place indeed.

> I had to add some logic to enforce the invariant that any fake offsets -> symbolic nested int entry in the registry has consistent ShapeEnv between the fake offsets and symbolic nested int. But maybe you've sidestepped this via your FakeTensor / FakeTensorMode changes?

That is interesting, thanks for the pointer. Though on the surface it looks like in `test_jagged_fake_to_fake_preserved` only a single shapeEnv is created (will dig into this more). Also, any idea how this can happen in practice?

soulitzer commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -622,13 +622,16 @@ def inner(*flat_args):
-        def view_avoid_dupes_with_primals(t):
+        def view_avoid_dupes_with_primals(t, parent_cls, attr):
             if isinstance(t, Tensor) and is_traceable_wrapper_subclass(t):
                 return transform_subclass(
-                    t, lambda _, inner_t: view_avoid_dupes_with_primals(inner_t)
+                    t, lambda attr, inner_t: view_avoid_dupes_with_primals(inner_t, type(t), attr)
                 )
             if isinstance(t, Tensor):
-                return t.view(t.shape)
+                out = t.view(t.shape)
+                if parent_cls is torch.nested._internal.nested_tensor.NestedTensor and (attr == "_offsets" or attr == "_lengths"):
+                    out.set_nested_int(t.nested_int())
+                return out
```
Nice clean up!

jbschlosser commented:
> That is interesting, thanks for the pointer. Though on the surface it looks like in test_jagged_fake_to_fake_preserved only a single shapeEnv is created (will dig into this more). Also, any idea how this can happen in practice?

Imagine running that test with `PYTORCH_TEST_WITH_DYNAMO=1` :) Now you've got the ShapeEnv in the test and the ShapeEnv used during tracing.

jbschlosser commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -622,13 +622,16 @@ def inner(*flat_args):
             if isinstance(t, Tensor):
-                return t.view(t.shape)
+                out = t.view(t.shape)
+                if parent_cls is torch.nested._internal.nested_tensor.NestedTensor and (attr == "_offsets" or attr == "_lengths"):
+                    out.set_nested_int(t.nested_int())
```
```suggestion
                if hasattr(t, "nested_int"):
                    out.set_nested_int(t.nested_int())
```

maybe? to keep it slightly cleaner

soulitzer committed:
Update on "Construct NJT without graph breaks"



Summary of changes (WIP):


**(1) Symbolic nested int are now cached on FakeTensor**

- The cache can be manually populated with `FakeTensor.set_nested_int()`

**(2) It is now the responsibility of the caller of `__tensor_unflatten__` to properly update the symbolic nested int  cache.**

- There are 4 cases in total today:
  - (i) initial fakification of the subclass
  - (ii) aot autograd taking a view of graph outputs to construct grad_output
  - (iii) aot autograd swapping FakeTensor with FunctionalTensor
  - (iv) aot autograd runtime wrapper

- We update case (i) and (ii) to update the FakeTensor cache. 
- If we want to avoid any nested specific logic at `__tensor_unflatten__` , we'd need to come up with some rule to decide whether to update. This rule is roughly "if previous inner tensor and new inner tensor have the same type" then we are okay to share their symbolic nested ints.

**(3) New trace-time specific tensor id counter.**

- Motivation: If we allow nested ints created within the traced region, we would like to avoid mutating the eager state. 

- This counter is set to the global tensor id counter upon entering fake mode, and reset upon exiting a fake mode, anticipating that a single fake mode can be used to trace the same program multiple times. 

- Piggybacking on the enter/exit of fake mode to reset the counter assumes that we don't enter/exit multiple times while tracing a single program. 

- In the case of a graph break, we should be fine if we execute the partial graph program before proceeding to trace the rest of the program.

~~**(4) Source is stored on FakeTensor for ALL tensors upon fakification**~~

- ~~`FakeTensor.nested_int()` must create a new symbol if not cached. The symbol's source is a ChainedSource based on the offset tensor's source. In order to have access to this source, this source is stored on the FakeTensor upon fakification. For intermediates that do not have source, we use an intermediate source. (unbacked symint does not immediately work here, maybe worth investigating more later).~~



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "Construct NJT without graph breaks"



Summary of changes (WIP):


**(1) Symbolic nested int are now cached on FakeTensor**

- The cache can be manually populated with `FakeTensor.set_nested_int()`

**(2) It is now the responsibility of the caller of `__tensor_unflatten__` to properly update the symbolic nested int  cache.**

- There are 4 cases in total today:
  - (i) initial fakification of the subclass
  - (ii) aot autograd taking a view of graph outputs to construct grad_output
  - (iii) aot autograd swapping FakeTensor with FunctionalTensor
  - (iv) aot autograd runtime wrapper

- We update case (i) and (ii) to update the FakeTensor cache. 
- If we want to avoid any nested specific logic at `__tensor_unflatten__` , we'd need to come up with some rule to decide whether to update. This rule is roughly "if previous inner tensor and new inner tensor have the same type" then we are okay to share their symbolic nested ints.

**(3) New trace-time specific tensor id counter.**

- Motivation: If we allow nested ints created within the traced region, we would like to avoid mutating the eager state. 

- This counter is set to the global tensor id counter upon entering fake mode, and reset upon exiting a fake mode, anticipating that a single fake mode can be used to trace the same program multiple times. 

- Piggybacking on the enter/exit of fake mode to reset the counter assumes that we don't enter/exit multiple times while tracing a single program. 

- In the case of a graph break, we should be fine if we execute the partial graph program before proceeding to trace the rest of the program.

~~**(4) Source is stored on FakeTensor for ALL tensors upon fakification**~~

- ~~`FakeTensor.nested_int()` must create a new symbol if not cached. The symbol's source is a ChainedSource based on the offset tensor's source. In order to have access to this source, this source is stored on the FakeTensor upon fakification. For intermediates that do not have source, we use an intermediate source. (unbacked symint does not immediately work here, maybe worth investigating more later).~~



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "Construct NJT without graph breaks"



Summary of changes (WIP):


**(1) Symbolic nested int are now cached on FakeTensor**

- The cache can be manually populated with `FakeTensor.set_nested_int()`

**(2) It is now the responsibility of the caller of `__tensor_unflatten__` to properly update the symbolic nested int  cache.**

- There are 4 cases in total today:
  - (i) initial fakification of the subclass
  - (ii) aot autograd taking a view of graph outputs to construct grad_output
  - (iii) aot autograd swapping FakeTensor with FunctionalTensor
  - (iv) aot autograd runtime wrapper

- We update case (i) and (ii) to update the FakeTensor cache. 
- If we want to avoid any nested specific logic at `__tensor_unflatten__` , we'd need to come up with some rule to decide whether to update. This rule is roughly "if previous inner tensor and new inner tensor have the same type" then we are okay to share their symbolic nested ints.

**(3) New trace-time specific tensor id counter.**

- Motivation: If we allow nested ints created within the traced region, we would like to avoid mutating the eager state. 

- This counter is set to the global tensor id counter upon entering fake mode, and reset upon exiting a fake mode, anticipating that a single fake mode can be used to trace the same program multiple times. 

- Piggybacking on the enter/exit of fake mode to reset the counter assumes that we don't enter/exit multiple times while tracing a single program. 

- In the case of a graph break, we should be fine if we execute the partial graph program before proceeding to trace the rest of the program.

~~**(4) Source is stored on FakeTensor for ALL tensors upon fakification**~~

- ~~`FakeTensor.nested_int()` must create a new symbol if not cached. The symbol's source is a ChainedSource based on the offset tensor's source. In order to have access to this source, this source is stored on the FakeTensor upon fakification. For intermediates that do not have source, we use an intermediate source. (unbacked symint does not immediately work here, maybe worth investigating more later).~~



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "Construct NJT without graph breaks"



Summary of changes (WIP):


**(1) Symbolic nested int are now cached on FakeTensor**

- The cache can be manually populated with `FakeTensor.set_nested_int()`

**(2) It is now the responsibility of the caller of `__tensor_unflatten__` to properly update the symbolic nested int  cache.**

- There are 4 cases in total today:
  - (i) initial fakification of the subclass
  - (ii) aot autograd taking a view of graph outputs to construct grad_output
  - (iii) aot autograd swapping FakeTensor with FunctionalTensor
  - (iv) aot autograd runtime wrapper

- We update case (i) and (ii) to update the FakeTensor cache. 
- If we want to avoid any nested specific logic at `__tensor_unflatten__` , we'd need to come up with some rule to decide whether to update. This rule is roughly "if previous inner tensor and new inner tensor have the same type" then we are okay to share their symbolic nested ints.

**(3) New trace-time specific tensor id counter.**

- Motivation: If we allow nested ints created within the traced region, we would like to avoid mutating the eager state. 

- This counter is set to the global tensor id counter upon entering fake mode, and reset upon exiting a fake mode, anticipating that a single fake mode can be used to trace the same program multiple times. 

- Piggybacking on the enter/exit of fake mode to reset the counter assumes that we don't enter/exit multiple times while tracing a single program. 

- In the case of a graph break, we should be fine if we execute the partial graph program before proceeding to trace the rest of the program.

~~**(4) Source is stored on FakeTensor for ALL tensors upon fakification**~~

- ~~`FakeTensor.nested_int()` must create a new symbol if not cached. The symbol's source is a ChainedSource based on the offset tensor's source. In order to have access to this source, this source is stored on the FakeTensor upon fakification. For intermediates that do not have source, we use an intermediate source. (unbacked symint does not immediately work here, maybe worth investigating more later).~~



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "Construct NJT without graph breaks"



Summary of changes (WIP):


**(1) Symbolic nested int are now cached on FakeTensor**

- The cache can be manually populated with `FakeTensor.set_nested_int()`

**(2) It is now the responsibility of the caller of `__tensor_unflatten__` to properly update the symbolic nested int  cache.**

- There are 4 cases in total today:
  - (i) initial fakification of the subclass
  - (ii) aot autograd taking a view of graph outputs to construct grad_output
  - (iii) aot autograd swapping FakeTensor with FunctionalTensor
  - (iv) aot autograd runtime wrapper

- We update case (i) and (ii) to update the FakeTensor cache. 
- If we want to avoid any nested specific logic at `__tensor_unflatten__` , we'd need to come up with some rule to decide whether to update. This rule is roughly "if previous inner tensor and new inner tensor have the same type" then we are okay to share their symbolic nested ints.

**(3) New trace-time specific tensor id counter.**

- Motivation: If we allow nested ints created within the traced region, we would like to avoid mutating the eager state. 

- This counter is set to the global tensor id counter upon entering fake mode, and reset upon exiting a fake mode, anticipating that a single fake mode can be used to trace the same program multiple times. 

- Piggybacking on the enter/exit of fake mode to reset the counter assumes that we don't enter/exit multiple times while tracing a single program. 

- In the case of a graph break, we should be fine if we execute the partial graph program before proceeding to trace the rest of the program.

~~**(4) Source is stored on FakeTensor for ALL tensors upon fakification**~~

- ~~`FakeTensor.nested_int()` must create a new symbol if not cached. The symbol's source is a ChainedSource based on the offset tensor's source. In order to have access to this source, this source is stored on the FakeTensor upon fakification. For intermediates that do not have source, we use an intermediate source. (unbacked symint does not immediately work here, maybe worth investigating more later).~~



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "Construct NJT without graph breaks"



Combines contributions from https://github.com/pytorch/pytorch/pull/130505

Some useful context in the comment block starting from:
https://github.com/pytorch/pytorch/blob/84bfb5c1d759932b327c9dc315c408d056825f5f/test/dynamo/test_subclasses.py#L1667

Descr (WIP)

Changes in this PR 
- For each tensor fakified, check the nested int registry in eager, and eagerly symbolicize if that tensor has already been associated with nested int in eager.
- Adds a separate counter stored on FakeTensorMode as a fake analog to _tensor_id_counter (which keeps track of unique tensors).
- (refactor) Today, we store FakeTensor -> symbolic nested int in the global registry. With this PR, symbolic nested int is stored directly on the FakeTensor. (Eager still caches nested int in the registry, though we should avoid this at some point.)

Basically unchanged, but worth noting:
- `__tensor_unflatten__` is still responsible for determining whether we should cache for now. The logic is somewhat simplified.
- to_copy is still using the trick of updating two different tensors in the registry to point to the same nested int. This is kind of broken, but we try to leave it as is, and plan a better fix with the UnionFind stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "Construct NJT without graph breaks"



Combines contributions from https://github.com/pytorch/pytorch/pull/130505

Some useful context in the comment block starting from:
https://github.com/pytorch/pytorch/blob/84bfb5c1d759932b327c9dc315c408d056825f5f/test/dynamo/test_subclasses.py#L1667

Descr (WIP)

Changes in this PR 
- For each tensor fakified, check the nested int registry in eager, and eagerly symbolicize if that tensor has already been associated with nested int in eager.
- Adds a separate counter stored on FakeTensorMode as a fake analog to _tensor_id_counter (which keeps track of unique tensors).
- (refactor) Today, we store FakeTensor -> symbolic nested int in the global registry. With this PR, symbolic nested int is stored directly on the FakeTensor. (Eager still caches nested int in the registry, though we should avoid this at some point.)

Basically unchanged, but worth noting:
- `__tensor_unflatten__` is still responsible for determining whether we should cache for now. The logic is somewhat simplified.
- to_copy is still using the trick of updating two different tensors in the registry to point to the same nested int. This is kind of broken, but we try to leave it as is, and plan a better fix with the UnionFind stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

soulitzer committed:
Update on "Construct NJT without graph breaks"



Combines contributions from https://github.com/pytorch/pytorch/pull/130505

Some useful context in the large comment block starting from:
https://github.com/pytorch/pytorch/blob/84bfb5c1d759932b327c9dc315c408d056825f5f/test/dynamo/test_subclasses.py#L1667

Descr (WIP)

Changes in this PR 
- For each tensor fakified, check the nested int registry in eager, and eagerly symbolicize if that tensor has already been associated with nested int in eager.
- Adds a separate counter stored on FakeTensorMode as a fake analog to _tensor_id_counter (which keeps track of unique tensors). This counter is initialized to the global eager tensor id counter upon creation of the FakeTensorMode, and needs to be reset when the same FakeTensorMode is reused to trace again (in this PR, we piggyback on the epoch incrementing logic).
- (refactor) Today, we store FakeTensor -> symbolic nested int in the global registry. With this PR, symbolic nested int is stored directly on the FakeTensor. (Eager still caches nested int in the registry, though we should avoid this at some point.)

Basically unchanged, but worth noting:
- `__tensor_unflatten__` is still responsible for determining whether we should cache for now. The logic is somewhat simplified.
- to_copy is still using the trick of updating two different tensors in the registry to point to the same nested int. This is kind of broken, but we try to leave it as is, and plan a better fix with the UnionFind stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

YuqingJ commented:
@yuqingj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60406772).

jbschlosser commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -22,6 +29,21 @@ def get_tensor_symint(tensor, *, coeff=1):
     return tensor_symint
 
 
+# Utility for testing
```
nit: if it's really just used for testing, imo it should be defined in a test file

jbschlosser commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -252,18 +276,14 @@ def __tensor_unflatten__(inner_tensors: Dict, meta, outer_size, outer_stride):
-        # Note that we cannot simply check if is_fake(values) because
-        # during aot autograd, FunctionalTensors are not fake but hold
-        # symbolic sizes.
+        # Alternatively, we could make it the caller's responsibility to
+        # cache it. But this heuristic seems simple enough.
```
I'm so sad to see this still here lol. I really want `tensor_unflatten()` to be straightforward and need to know nothing about fake tensors, tracing, etc.

I think the nested int should be cached during fake tensor construction for the strong invariant: a fake tensor post-creation is guaranteed to have any associated symbolic nested int set. I think this is done for one path during fake-ification but the `_empty_create_subclass()` path relies on `tensor_unflatten()` to set the nested int? Imo the responsibility to set it should be in fake-ification only and not in the scope of `tensor_unflatten()`.

Also, this logic will become irrelevant after the nested int revamp, so why not get closer to that ideal now :p

soulitzer replied:
> a fake tensor post-creation is guaranteed to have any associated symbolic nested int set

I had this earlier but removed it to make this PR smaller since it seemed like an orthogonal change. I like the idea of having this invariant earlier though, that is a great way to put it. I'll bring it back in an earlier PR.

soulitzer replied:
Opened a follow up PR https://github.com/pytorch/pytorch/pull/132414 to do this.

jbschlosser commented on torch/_subclasses/meta_utils.py:
```diff
@@ -1571,6 +1579,12 @@ def is_c_of_r(complex_dtype, real_dtype):
                 r._is_param = True
 
+            # See Note: [Creating symbolic nested int]
+            if t.nested_int is not None and not r.has_nested_int():
```
hm how can `r.has_nested_int()` ever evaluate to `True` here? IIUC `r` was just created and this is the point at which we would set the symbolic nested int for it

jbschlosser reviewed (commented):
Thanks for all the hard work on this! It's crucial to get this in :)

Some nits but I'm good with it - I'll defer to Ed / Brian for the approval stamp though

soulitzer committed:
Update on "Construct NJT without graph breaks"



Combines contributions from https://github.com/pytorch/pytorch/pull/130505

Some context can be found in this large comment block:

https://github.com/pytorch/pytorch/blob/a5b64d39fd50d47dfe6f821941ce4c803c7c4d63/test/dynamo/test_subclasses.py#L1667-L1681

Changes in this PR 
- For each tensor fakified, check the nested int registry in eager, and eagerly symbolicize if that tensor has already been associated with nested int in eager.
- Adds a separate counter stored on FakeTensorMode as a fake analog to _tensor_id_counter (which keeps track of unique tensors). This counter is initialized to the global eager tensor id counter upon creation of the FakeTensorMode, and needs to be reset when the same FakeTensorMode is reused to trace again (in this PR, we piggyback on the epoch incrementing logic).
- (refactor) Today, we store FakeTensor -> symbolic nested int in the global registry. With this PR, symbolic nested int is stored directly on the FakeTensor. (Eager still caches nested int in the registry, though we should avoid this at some point.)

Basically unchanged, but worth noting:
- `__tensor_unflatten__` is still responsible for determining whether we should cache for now. The logic is somewhat simplified.
- to_copy is still using the trick of updating two different tensors in the registry to point to the same nested int. This is kind of broken, but we try to leave it as is, and plan a better fix with the UnionFind stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

Differential Revision: [D60406772](https://our.internmc.facebook.com/intern/diff/D60406772)

[ghstack-poisoned]

soulitzer commented on torch/testing/_internal/common_utils.py:
```diff
@@ -5178,6 +5178,34 @@ def wrapped(self, *args, **kwargs):
+            else:
+                return x
+
+        self.assertEqual(pytree.tree_map(_unbind_njts, a), pytree.tree_map(_unbind_njts, b))
```
Just code motion so that it can be reused more easily

soulitzer commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -587,6 +587,9 @@ class FakeTensor(Tensor):
     unique_memo = UnbackedMemoDescriptor()
 
+    _nested_int_memo: Optional[torch.SymInt]
+    _nested_int_memo_vc: Optional[int]
```
Some adjustments are needed, but seems not too bad.  1) Made it more generic rather than unbacked specific, 2) I want to be able to operate generically on both FakeTensor and FunctionalTensor so had to add a "PassThroughDescriptor" there.

soulitzer committed:
Update on "Construct NJT without graph breaks"



Combines contributions from https://github.com/pytorch/pytorch/pull/130505

Some context can be found in this large comment block:

https://github.com/pytorch/pytorch/blob/a5b64d39fd50d47dfe6f821941ce4c803c7c4d63/test/dynamo/test_subclasses.py#L1667-L1681

Changes in this PR 
- For each tensor fakified, check the nested int registry in eager, and eagerly symbolicize if that tensor has already been associated with nested int in eager.
- Adds a separate counter stored on FakeTensorMode as a fake analog to _tensor_id_counter (which keeps track of unique tensors). This counter is initialized to the global eager tensor id counter upon creation of the FakeTensorMode, and needs to be reset when the same FakeTensorMode is reused to trace again (in this PR, we piggyback on the epoch incrementing logic).
- (refactor) Today, we store FakeTensor -> symbolic nested int in the global registry. With this PR, symbolic nested int is stored directly on the FakeTensor. (Eager still caches nested int in the registry, though we should avoid this at some point.)

Basically unchanged, but worth noting:
- `__tensor_unflatten__` is still responsible for determining whether we should cache for now. The logic is somewhat simplified.
- to_copy is still using the trick of updating two different tensors in the registry to point to the same nested int. This is kind of broken, but we try to leave it as is, and plan a better fix with the UnionFind stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

Differential Revision: [D60406772](https://our.internmc.facebook.com/intern/diff/D60406772)

[ghstack-poisoned]

bdhirsh commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -860,6 +867,15 @@ def merge_devices(t: object) -> None:
+        *,
+        coeff: Union[int, torch.SymInt] = 1,
+    ) -> torch.SymInt:
+        if self.nested_int_memo is None:
```
do we ever want this to be None?

soulitzer replied:
We expect it to be None when an offsets is a graph intermediate, or an input that has never been associated with a nested int, and then that offsets is later used during NT construction.

bdhirsh replied:
do you mind adding that explanation in a comment near where the memo is defined?

bdhirsh commented on torch/nested/_internal/ops.py:
```diff
@@ -472,7 +472,15 @@ def to_copy_default(func, *args, **kwargs):
+
+    if isinstance(new_offsets, (FakeTensor, FunctionalTensor)):
+        # Temporary hack until we have the union find
+        new_offsets.nested_int_memo = inp._offsets.nested_int_memo
```
I was just looking around trying to understand why we need to define a `nested_into_memo` and `get_nested_int` in `FunctionalTensor` directly - is this the reason?

instead of having FunctionalTensor know about nested_int directly, I think I would rather have the nested code that needs to get/set a `nested_int` just know about FunctionalTensor and do any necessary unwrapping (it also needs to know about `FunctionalTensor` in the isinstance check above). Does that seem reasonable?

soulitzer replied:
It actually looks like the aot autograd calling view_as to produce grad_outputs logic does not run with FunctionalTensors. Updated here and the constructor to unwrap as needed.

bdhirsh commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -252,18 +260,14 @@ def __tensor_unflatten__(inner_tensors: Dict, meta, outer_size, outer_stride):
+        if isinstance(ragged_source, FakeTensor):
             ragged_size = outer_size[ragged_idx]
-            _tensor_symint_registry[ragged_source] = ragged_size
+            ragged_source.nested_int_memo = ragged_size
```
I think `ragged_source can be a `FunctionalTensor(FakeTensor)` here as well. Is it intentional that we are only setting the memo in the FakeTensor case?

(just thinking out loud - I think the answer is... yes, because you've written code to make `FunctionalTensor.nested_into_memo` a pass-through to grab the nested_int from the underlying FakeTensor. But might be worth calling out the invariant in a comment somewhere, that only the `FakeTensor` offsets gets a memo)

soulitzer replied:
Yes, will add a comment

soulitzer replied:
This particular piece of code actually goes away (see next PR), so I added the comment to somewhere else (get_nested_int logic).

bdhirsh commented on torch/_subclasses/meta_utils.py:
```diff
@@ -458,6 +461,10 @@ class MetaTensorDesc:
+    # We eagerly symbolicize the associated nested int for e.g. offsets / lengths
+    # metadata if that offsets is already associated with a nested int.
+    # See test_construct_from_jagged_with_input_offsets_mixed_case.
+    nested_int: Optional[int] = None
```
hmm... `Optional[int]` is wrong, right? (since this is a nested_int). But `nested_int` is a type of `SymInt`, and... putting `SymInt` here feels wrong since `MetaTensorDesc` is serializable but SymInts in general are not serializable.

Do we have another python type to tell nested ints apart from other SymInts? Or maybe SymInt is fine here just for type annotations

soulitzer replied:
I think the idea here is that we imagine nested int to be some POD uniquely identified by its "id", so we don't necessarily need to store a particular instance.

soulitzer replied:
(You were right in that we were actually saving the nested int object today, I've just updated to actually use an int)

bdhirsh commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -2100,6 +2138,30 @@ def wrap(e: T) -> Union[T, FakeTensor]:
         return tree_map(wrap, r)
 
+    def create_symbolic_nested_int(
+        self, *, hint: Optional[Union[int, torch.SymInt]] = None
```
similar q, but what is the invariant around the type of hint if it is not None (can it really be either any int/symint?). Or is it guaranteed to be a nested_int, and we just don't have a better python type annotation for it

soulitzer replied:
I'm expecting the hint to always be a plain nested int. So the annotation should be a `Optional[torch.SymInt]` and this might just be copypasta.

soulitzer replied:
Actually this was related to https://github.com/pytorch/pytorch/pull/130292#discussion_r1702041479, and this should be an `Optional[int]`!

bdhirsh commented on torch/fx/passes/fake_tensor_prop.py:
```diff
@@ -30,6 +30,7 @@ def __init__(self, module: torch.fx.GraphModule, mode: Optional[FakeTensorMode]
             mode = FakeTensorMode()
         self._mode = mode
         mode.epoch += 1
+        mode.reset_nt_tensor_id_counter()
```
just thinking: would it be correct / less error prone just to reset this directly in `FakeTensorMode.__enter__`, so we don't have to worry about doing it manually?

soulitzer replied:
This is a good question. - I do remember trying this earlier and I logging when enter/exit would be called, it appears that we enter/exit every time we dispatch into any op in FakeTensorMode's torch dispatch. Would need to double check if there is a outer context manager that is still enabled while all of this is happening (maybe it is fine to rely on enter/exit if that is the case?)

soulitzer replied:
Okay, I added a breakpoint to `__enter__`'s  is a no-op path, and it is never reached. I'm not sure we can rely on enter/exit to do the right thing.

bdhirsh replied:
yeah... i think the fake tensor caching logic re-enables/disables the fake mode so maybe this is a pain :/

bdhirsh commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -13,7 +12,14 @@
+    from torch._subclasses.functional_tensor import FunctionalTensor
+
+    if isinstance(tensor, (FakeTensor, FunctionalTensor)):
+        return tensor.get_nested_int(coeff=coeff)
```
confirming my understanding, but:

(1) one invariant we want is that the global `_tensor_symint_registry` is never mutated during tracing (and only contains real tracing)
(2) that invariant is maintained here via the short-circuit
(3) we get "reasonable" behavior today because at runtime, any offset tensors that become inputs to NJT constructors in the graph (that were previously not associated with an NJT) will get added to the dict here in AOTDispatcher's runtime wrapper on exit (when we wrap NJT tensor subclass graph outputs)

soulitzer replied:
Yep, that is correct

bdhirsh commented:
looks ok to me, happy to stamp and/or defer to Ed

soulitzer committed:
Update on "Construct NJT without graph breaks"



Combines contributions from https://github.com/pytorch/pytorch/pull/130505

Some context can be found in this large comment block:

https://github.com/pytorch/pytorch/blob/a5b64d39fd50d47dfe6f821941ce4c803c7c4d63/test/dynamo/test_subclasses.py#L1667-L1681

Changes in this PR 
- For each tensor fakified, check the nested int registry in eager, and eagerly symbolicize if that tensor has already been associated with nested int in eager.
- Adds a separate counter stored on FakeTensorMode as a fake analog to _tensor_id_counter (which keeps track of unique tensors). This counter is initialized to the global eager tensor id counter upon creation of the FakeTensorMode, and needs to be reset when the same FakeTensorMode is reused to trace again (in this PR, we piggyback on the epoch incrementing logic).
- (refactor) Today, we store FakeTensor -> symbolic nested int in the global registry. With this PR, symbolic nested int is stored directly on the FakeTensor. (Eager still caches nested int in the registry, though we should avoid this at some point.)

Basically unchanged, but worth noting:
- `__tensor_unflatten__` is still responsible for determining whether we should cache for now. The logic is somewhat simplified.
- to_copy is still using the trick of updating two different tensors in the registry to point to the same nested int. This is kind of broken, but we try to leave it as is, and plan a better fix with the UnionFind stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

Differential Revision: [D60406772](https://our.internmc.facebook.com/intern/diff/D60406772)

[ghstack-poisoned]

soulitzer committed:
Update on "Construct NJT without graph breaks"



Combines contributions from https://github.com/pytorch/pytorch/pull/130505

Some context can be found in this large comment block:

https://github.com/pytorch/pytorch/blob/a5b64d39fd50d47dfe6f821941ce4c803c7c4d63/test/dynamo/test_subclasses.py#L1667-L1681

Changes in this PR 
- For each tensor fakified, check the nested int registry in eager, and eagerly symbolicize if that tensor has already been associated with nested int in eager.
- Adds a separate counter stored on FakeTensorMode as a fake analog to _tensor_id_counter (which keeps track of unique tensors). This counter is initialized to the global eager tensor id counter upon creation of the FakeTensorMode, and needs to be reset when the same FakeTensorMode is reused to trace again (in this PR, we piggyback on the epoch incrementing logic).
- (refactor) Today, we store FakeTensor -> symbolic nested int in the global registry. With this PR, symbolic nested int is stored directly on the FakeTensor. (Eager still caches nested int in the registry, though we should avoid this at some point.)

Basically unchanged, but worth noting:
- `__tensor_unflatten__` is still responsible for determining whether we should cache for now. The logic is somewhat simplified.
- to_copy is still using the trick of updating two different tensors in the registry to point to the same nested int. This is kind of broken, but we try to leave it as is, and plan a better fix with the UnionFind stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

Differential Revision: [D60406772](https://our.internmc.facebook.com/intern/diff/D60406772)

[ghstack-poisoned]

soulitzer committed:
Update on "Construct NJT without graph breaks"



Combines contributions from https://github.com/pytorch/pytorch/pull/130505

Some context can be found in this large comment block:

https://github.com/pytorch/pytorch/blob/a5b64d39fd50d47dfe6f821941ce4c803c7c4d63/test/dynamo/test_subclasses.py#L1667-L1681

Changes in this PR 
- For each tensor fakified, check the nested int registry in eager, and eagerly symbolicize if that tensor has already been associated with nested int in eager.
- Adds a separate counter stored on FakeTensorMode as a fake analog to _tensor_id_counter (which keeps track of unique tensors). This counter is initialized to the global eager tensor id counter upon creation of the FakeTensorMode, and needs to be reset when the same FakeTensorMode is reused to trace again (in this PR, we piggyback on the epoch incrementing logic).
- (refactor) Today, we store FakeTensor -> symbolic nested int in the global registry. With this PR, symbolic nested int is stored directly on the FakeTensor. (Eager still caches nested int in the registry, though we should avoid this at some point.)

Basically unchanged, but worth noting:
- `__tensor_unflatten__` is still responsible for determining whether we should cache for now. The logic is somewhat simplified.
- to_copy is still using the trick of updating two different tensors in the registry to point to the same nested int. This is kind of broken, but we try to leave it as is, and plan a better fix with the UnionFind stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

Differential Revision: [D60406772](https://our.internmc.facebook.com/intern/diff/D60406772)

[ghstack-poisoned]

ezyang commented:
Deferring to Brian here

soulitzer committed:
Update on "Construct NJT without graph breaks"



Combines contributions from https://github.com/pytorch/pytorch/pull/130505

Some context can be found in this large comment block:

https://github.com/pytorch/pytorch/blob/a5b64d39fd50d47dfe6f821941ce4c803c7c4d63/test/dynamo/test_subclasses.py#L1667-L1681

Changes in this PR 
- For each tensor fakified, check the nested int registry in eager, and eagerly symbolicize if that tensor has already been associated with nested int in eager.
- Adds a separate counter stored on FakeTensorMode as a fake analog to _tensor_id_counter (which keeps track of unique tensors). This counter is initialized to the global eager tensor id counter upon creation of the FakeTensorMode, and needs to be reset when the same FakeTensorMode is reused to trace again (in this PR, we piggyback on the epoch incrementing logic).
- (refactor) Today, we store FakeTensor -> symbolic nested int in the global registry. With this PR, symbolic nested int is stored directly on the FakeTensor. (Eager still caches nested int in the registry, though we should avoid this at some point.)

Basically unchanged, but worth noting:
- `__tensor_unflatten__` is still responsible for determining whether we should cache for now. The logic is somewhat simplified.
- to_copy is still using the trick of updating two different tensors in the registry to point to the same nested int. This is kind of broken, but we try to leave it as is, and plan a better fix with the UnionFind stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D60406772](https://our.internmc.facebook.com/intern/diff/D60406772)

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge

soulitzer committed:
Update on "Construct NJT without graph breaks"



Combines contributions from https://github.com/pytorch/pytorch/pull/130505

Some context can be found in this large comment block:

https://github.com/pytorch/pytorch/blob/a5b64d39fd50d47dfe6f821941ce4c803c7c4d63/test/dynamo/test_subclasses.py#L1667-L1681

Changes in this PR 
- For each tensor fakified, check the nested int registry in eager, and eagerly symbolicize if that tensor has already been associated with nested int in eager.
- Adds a separate counter stored on FakeTensorMode as a fake analog to _tensor_id_counter (which keeps track of unique tensors). This counter is initialized to the global eager tensor id counter upon creation of the FakeTensorMode, and needs to be reset when the same FakeTensorMode is reused to trace again (in this PR, we piggyback on the epoch incrementing logic).
- (refactor) Today, we store FakeTensor -> symbolic nested int in the global registry. With this PR, symbolic nested int is stored directly on the FakeTensor. (Eager still caches nested int in the registry, though we should avoid this at some point.)

Basically unchanged, but worth noting:
- `__tensor_unflatten__` is still responsible for determining whether we should cache for now. The logic is somewhat simplified.
- to_copy is still using the trick of updating two different tensors in the registry to point to the same nested int. This is kind of broken, but we try to leave it as is, and plan a better fix with the UnionFind stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D60406772](https://our.internmc.facebook.com/intern/diff/D60406772)

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge

soulitzer committed:
Update on "Construct NJT without graph breaks"



Combines contributions from https://github.com/pytorch/pytorch/pull/130505

Some context can be found in this large comment block:

https://github.com/pytorch/pytorch/blob/a5b64d39fd50d47dfe6f821941ce4c803c7c4d63/test/dynamo/test_subclasses.py#L1667-L1681

Changes in this PR 
- For each tensor fakified, check the nested int registry in eager, and eagerly symbolicize if that tensor has already been associated with nested int in eager.
- Adds a separate counter stored on FakeTensorMode as a fake analog to _tensor_id_counter (which keeps track of unique tensors). This counter is initialized to the global eager tensor id counter upon creation of the FakeTensorMode, and needs to be reset when the same FakeTensorMode is reused to trace again (in this PR, we piggyback on the epoch incrementing logic).
- (refactor) Today, we store FakeTensor -> symbolic nested int in the global registry. With this PR, symbolic nested int is stored directly on the FakeTensor. (Eager still caches nested int in the registry, though we should avoid this at some point.)

Basically unchanged, but worth noting:
- `__tensor_unflatten__` is still responsible for determining whether we should cache for now. The logic is somewhat simplified.
- to_copy is still using the trick of updating two different tensors in the registry to point to the same nested int. This is kind of broken, but we try to leave it as is, and plan a better fix with the UnionFind stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D60406772](https://our.internmc.facebook.com/intern/diff/D60406772)

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge

soulitzer committed:
Update on "Construct NJT without graph breaks"



Combines contributions from https://github.com/pytorch/pytorch/pull/130505

Some context can be found in this large comment block:

https://github.com/pytorch/pytorch/blob/a5b64d39fd50d47dfe6f821941ce4c803c7c4d63/test/dynamo/test_subclasses.py#L1667-L1681

Changes in this PR 
- For each tensor fakified, check the nested int registry in eager, and eagerly symbolicize if that tensor has already been associated with nested int in eager.
- Adds a separate counter stored on FakeTensorMode as a fake analog to _tensor_id_counter (which keeps track of unique tensors). This counter is initialized to the global eager tensor id counter upon creation of the FakeTensorMode, and needs to be reset when the same FakeTensorMode is reused to trace again (in this PR, we piggyback on the epoch incrementing logic).
- (refactor) Today, we store FakeTensor -> symbolic nested int in the global registry. With this PR, symbolic nested int is stored directly on the FakeTensor. (Eager still caches nested int in the registry, though we should avoid this at some point.)

Basically unchanged, but worth noting:
- `__tensor_unflatten__` is still responsible for determining whether we should cache for now. The logic is somewhat simplified.
- to_copy is still using the trick of updating two different tensors in the registry to point to the same nested int. This is kind of broken, but we try to leave it as is, and plan a better fix with the UnionFind stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D60406772](https://our.internmc.facebook.com/intern/diff/D60406772)

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge

soulitzer commented:
@pytorchbot merge -i

soulitzer commented:
@pytorchbot revert -r nosignal -m "Causing a sev"

soulitzer commented:
@pytorchbot revert -m "causing a sev" -c nosignal

soulitzer commented:
Doing the revert here instead https://github.com/pytorch/pytorch/pull/133145
